core/src/main/java/org/elasticsearch/search/aggregations/metrics/percentiles/hdr/AbstractInternalHDRPercentiles.java

```diff
@@ -0,0 +1,145 @@
+import org.elasticsearch.search.aggregations.metrics.InternalNumericMetricsAggregation;
+import org.elasticsearch.search.aggregations.pipeline.PipelineAggregator;
+import org.elasticsearch.search.aggregations.support.format.ValueFormatter;
+import org.elasticsearch.search.aggregations.support.format.ValueFormatterStreams;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.Map;
+import java.util.zip.DataFormatException;
+
+abstract class AbstractInternalHDRPercentiles extends InternalNumericMetricsAggregation.MultiValue {
+
+    protected double[] keys;
+    protected DoubleHistogram state;
+    private boolean keyed;
+
+    AbstractInternalHDRPercentiles() {} // for serialization
+
+    public AbstractInternalHDRPercentiles(String name, double[] keys, DoubleHistogram state, boolean keyed, ValueFormatter formatter,
+            List<PipelineAggregator> pipelineAggregators,
+            Map<String, Object> metaData) {
+        super(name, pipelineAggregators, metaData);
+        this.keys = keys;
+        this.state = state;
+        this.keyed = keyed;
+        this.valueFormatter = formatter;
+    }
+
+    @Override
+    public double value(String name) {
+        return value(Double.parseDouble(name));
+    }
+
+    public abstract double value(double key);
+
+    public long getEstimatedMemoryFootprint() {
+        return state.getEstimatedFootprintInBytes();
+    }
+
+    @Override
+    public AbstractInternalHDRPercentiles doReduce(List<InternalAggregation> aggregations, ReduceContext reduceContext) {
+        DoubleHistogram merged = null;
+        for (InternalAggregation aggregation : aggregations) {
+            final AbstractInternalHDRPercentiles percentiles = (AbstractInternalHDRPercentiles) aggregation;
+            if (merged == null) {
+                merged = new DoubleHistogram(percentiles.state);
+                merged.setAutoResize(true);
+            }
+            merged.add(percentiles.state);
+        }
+        return createReduced(getName(), keys, merged, keyed, pipelineAggregators(), getMetaData());
+    }
+
+    protected abstract AbstractInternalHDRPercentiles createReduced(String name, double[] keys, DoubleHistogram merged, boolean keyed,
+            List<PipelineAggregator> pipelineAggregators, Map<String, Object> metaData);
+
+    @Override
+    protected void doReadFrom(StreamInput in) throws IOException {
+        valueFormatter = ValueFormatterStreams.readOptional(in);
+        keys = new double[in.readInt()];
+        for (int i = 0; i < keys.length; ++i) {
+            keys[i] = in.readDouble();
+        }
+        long minBarForHighestToLowestValueRatio = in.readLong();
+        ByteBuffer stateBuffer = ByteBuffer.wrap(in.readByteArray());
+        try {
+            state = DoubleHistogram.decodeFromCompressedByteBuffer(stateBuffer, minBarForHighestToLowestValueRatio);
+            state.setAutoResize(true);
```

Comment:
Elasticsearch guarantees that two nodes that have the same major version can talk to each other, so I'm wondering if HDRHistogram has a similar warranty so that upgrading HDRHistogram in a minor Elasticsearch release would not break Elasticsearch's wire protocol.